src/connectionController.ts

```diff
@@ -116,142 +122,188 @@ export default class ConnectionController {
-        secrets
+          ...savedConnectionInfo,
+          secretStorageLocation: SecretStorageLocation.SecretStorage,
+        },
+        keytarSecrets
       );
 
-      return {
-        ...savedConnectionInfo,
-        connectionOptions: connectionInfoWithSecrets.connectionOptions,
-      };
-    } catch (error) {
-      // Here we're lenient when loading connections in case their
-      // connections have become corrupted.
-      log.error('Getting connection info with secrets failed', error);
-      return;
-    }
-  }
-
-  private async _loadSavedConnectionsByStore(
-    savedConnections: ConnectionsFromStorage
-  ): Promise<void> {
-    if (!savedConnections || !Object.keys(savedConnections).length) {
-      return;
-    }
-
-    /** User connections are being saved both in:
-     * 1. Vscode global/workspace storage (without secrets) + keychain (secrets)
-     * 2. Memory of the extension (with secrets)
-     */
-    await Promise.all(
-      Object.keys(savedConnections).map(async (connectionId) => {
-        // Get connection info from vscode storage and merge with secrets.
-        const connectionInfoWithSecrets =
-          await this._getConnectionInfoWithSecrets(
-            savedConnections[connectionId]
-          );
-
-        // Save connection info with secrets to extension memory.
-        if (connectionInfoWithSecrets) {
-          this._connections[connectionId] = connectionInfoWithSecrets;
-        }
+    await this._saveConnectionWithSecret(migratedConnectionInfoWithSecrets);
 
-        this.eventEmitter.emit(DataServiceEventTypes.CONNECTIONS_DID_CHANGE);
-      })
-    );
+    return migratedConnectionInfoWithSecrets;
   }
 
-  async loadSavedConnections(): Promise<void> {
-    await Promise.all([
-      (async () => {
-        // Try to pull in the connections previously saved in the global storage of vscode.
-        const existingGlobalConnections = this._storageController.get(
-          StorageVariables.GLOBAL_SAVED_CONNECTIONS,
-          StorageLocation.GLOBAL
-        );
-        await this._loadSavedConnectionsByStore(existingGlobalConnections);
-      })(),
-      (async () => {
-        // Try to pull in the connections previously saved in the workspace storage of vscode.
-        const existingWorkspaceConnections = this._storageController.get(
-          StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
-          StorageLocation.WORKSPACE
-        );
-        await this._loadSavedConnectionsByStore(existingWorkspaceConnections);
-      })(),
-    ]);
+  _mergedConnectionInfoWithSecrets<
+    T extends StoreConnectionInfoWithConnectionOptions
+  >(connectionInfo: T, unparsedSecrets: string) {
+    if (!unparsedSecrets) {
+      return connectionInfo;
+    }
+
+    const secrets = JSON.parse(unparsedSecrets);
+    const connectionInfoWithSecrets = mergeSecrets(
+      {
+        id: connectionInfo.id,
+        connectionOptions: connectionInfo.connectionOptions,
+      } as ConnectionInfo,
```

Comment:
I realize this is existing code, but 
```suggestion
      },
```

compiles for me, so unless I’m missing something, we should remove the cast here :slightly_smiling_face: